File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def replacement(data) :
    data = data.replace('C#', 'c')
    data = data.replace('D#', 'd')
    data = data.replace('F#', 'f')
    data = data.replace('G#', 'g')
    data = data.replace('A#', 'a')

    return data

def solution(m, musicinfos) :
    answer = ''
    m = replacement(m)

    max_time = 0
    for music in musicinfos :
        start, end, name, info = music.split(",")
        info = replacement(info)

        play_time = (int(end[:2]) - int(start[:2])) * 60 + (int(end[3:]) - int(start[3:]))
        index = 0
        value = ''
        time = play_time
        while time :
            value += info[index]
            if index + 1 == len(info) :
                index = -1
            index += 1
            time -= 1

        if m in value :
            if answer :
                if max_time >= play_time :
                    continue
            max_time = play_time
            answer = name

    if answer :
        return answer
    else :
        return '(None)'

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.js
function solution(m, musicinfos) {
    const normalize = (melody) => {
        return melody.replace(/([A-Z])#/g, (_, group) => group.toLowerCase());
    };
    
    const getMin = (time) => {
        const [hour, minute] = time.split(':').map(Number);
        return hour * 60 + minute;
    };
    
    const getDiff = (start, end) => {
        return getMin(end) - getMin(start);
    };
    
    const music = musicinfos.map(info => {
        const [start, end, title, melody] = info.split(',');
        return {
            time: getDiff(start, end),
            title,
            melody: normalize(melody)
        };
    });
    
    const targetMelody = normalize(m);
    let output = { title: '(None)' };
    music.forEach(info => {
        const { time, title, melody } = info;
        let compMelody = melody.repeat(2);
        let count = 2;
        while (targetMelody.length + melody.length - 1 > compMelody.length) {
            compMelody += melody;
            count++;
        }
        const index = compMelody.indexOf(targetMelody);
        if (index >= 0 && index + targetMelody.length <= time) {
            if (output.title === '(None)' || output.time < time) {
                output = info;
            }
        }
    });
    return output.title;
}